### PR TITLE
fix: raise not implemented error when REST transport is not supported

### PR DESCRIPTION
**Pull request by parthea**
Fixes b/266705376

### CODE REVIEW

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2 [hunk 1]
```diff
@@ -428,8 +428,7 @@ class {{service.name}}RestTransport({{service.name}}Transport):
             {% endif %}{# method.void #}
      {% else %}{# method.http_options and not method.client_streaming #}
             {% if not method.http_options %}
-            raise RuntimeError(
-                "Cannot define a method without a valid 'google.api.http' annotation.")
+            raise NotImplementedError("This method does not support REST transport.")
```
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:431]: On my local branch, I was simplifying this whole else clause to also account for the streaming error. I think that's worth doing:
```
     {% else %}{# method.http_options and not method.client_streaming #}
            raise NotImplementedError(
                "Method {{ method.name }} is not available over REST transport")

    {%- endif %}{# unary method #}

```

**Review by vchudnov-g — COMMENTED** [review]
Looking good. One sugestion.

Also a question: how did you invoke the tests? I had problems getting the tests to fail after I changed the code but not the test files.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2 (hunk 1 above)
Comment (parthea) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2]: Done.

**Comment by parthea** [comment]
> Also a question: how did you invoke the tests? I had problems getting the tests to fail after I changed the code but not the test files.

I ran `nox -s fragment-3.9` and saw tests fail


```
(py39) partheniou@partheniou-vm-2:~/git/gapic-generator-python$ nox -s fragment-3.9
nox > Running session fragment-3.9
nox > Creating virtual environment (virtualenv) using python3.9 in .nox/fragment-3-9
nox > python -m pip install coverage pytest pytest-cov pytest-xdist 'asyncmock; python_version < '"'"'3.8'"'"'' pytest-asyncio grpcio-tools
nox > python -m pip install -e .
nox > python -m grpc_tools.protoc --proto_path=tests/fragments --python_gapic_out=/tmp/tmpr_8183ox --python_gapic_opt=transport=grpc+rest,python-gapic-templates=DEFAULT test_recursive_messages.proto
nox > python -m pip install /tmp/tmpr_8183ox -e . -t /tmp/tmpr_8183ox -qqq
nox > py.test --quiet --cov-config=/tmp/tmpr_8183ox/.coveragerc --cov-report=term --cov-fail-under=100 /tmp/tmpr_8183ox/tests/unit
nox > Command py.test --quiet --cov-config=/tmp/tmpr_8183ox/.coveragerc --cov-report=term --cov-fail-under=100 /tmp/tmpr_8183ox/tests/unit failed with exit code 1:
.....................................F.................................. [ 79%]
...................                                                      [100%]
=================================== FAILURES ===================================
__________________________ test_my_method_rest_error ___________________________

    def test_my_method_rest_error():
        client = MyServiceClient(
            credentials=ga_credentials.AnonymousCredentials(),
            transport='rest'
        )
        # Since a `google.api.http` annotation is required for using a rest transport
        # method, this should error.
        with pytest.raises(RuntimeError) as runtime_error:
            client.my_method({})
>       assert ("Cannot define a method without a valid 'google.api.http' annotation."
            in str(runtime_error.value))
E       assert "Cannot define a method without a valid 'google.api.http' annotation." in 'Method MyMethod is not available over REST transport'
E        +  where 'Method MyMethod is not available over REST transport' = str(NotImplementedError('Method MyMethod is not available over REST transport'))
E        +    where NotImplementedError('Method MyMethod is not available over REST transport') = <ExceptionInfo NotImplementedError('Method MyMethod is not available over REST transport') tblen=5>.value

/tmp/tmpr_8183ox/tests/unit/gapic/fragment/test_my_service.py:646: AssertionError
=========================== short test summary info ============================
FAILED ../../../../../../../tmp/tmpr_8183ox/tests/unit/gapic/fragment/test_my_service.py::test_my_method_rest_error
1 failed, 90 passed in 1.05s
nox > Session fragment-3.9 failed.
```

**Review by vchudnov-g — COMMENTED** [review]
Looks good, but we should make the equivalent changes for `ads-templates/` to prevent technical debt.

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2 [hunk 2]
```diff
@@ -642,10 +642,12 @@ def test_{{ method_name }}_rest_error():
...
     # method, this should error.
-    with pytest.raises(RuntimeError) as runtime_error:
+    with pytest.raises(NotImplementedError) as not_implemented_error:
         client.{{ method_name }}({})
-    assert ("Cannot define a method without a valid 'google.api.http' annotation."
-        in str(runtime_error.value))
+    assert (
+        "Method {{ method.name }} is not available over REST transport"
+        in str(not_implemented_error.value)
+    )
```
Comment (atulep) [inline gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:650]: Shouldn't this be indented forward - i.e. same indent as `client.{{ method_name }}` ?

Comment (vchudnov-g) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:650]: I would have thought so too, but the `not_implemented_error` remains in scope after the `with`, and the `with` catches the exception. See also https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

**Review by vchudnov-g — APPROVED** [review]
Looks great! Thanks for doing this.

**Pull request merged**